### ISSUE TITLE
Support rare/undocumented features of Flow

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarJS#771]
Relates to #584 
- [ ] `declare module "foo" { declare export function bar(p1: number): string; }`
- [ ] `declare module "M" { import type T from "TM"; }`
- [ ] Declare class mixins `declare class A mixins B<T>, C {}`
- [ ] `class C<+T,-U> {}`
- [ ] `declare export opaque type Foo`, `declare opaque type Test: Foo`
- [ ] %checks `var f = (x: mixed): %checks => typeof x === "string”;`
- [x] empty tuples `var a : [] = [];` #811
- [ ] Rest property in object type, interface
- [ ] `const x = 1` as object/interface property
- [x] `export type { Foo, Bar }` export with unnamed type #833 
- [x] `export type {foo} from 'bar'` #833

**Comment by marcdonovan** [comment]
```
10:33:14.811 ERROR: Parse error at line 4 column 13:

4: export type {
                ^
5:   ErrorReport, 
6: };
```

Is this a rare/undocumented?

**Comment by vilchik-elena** [comment]
@marcdonovan thanks for letting us know! Added it to the list of not supported syntax in this issue

**Comment by marcdonovan** [comment]
getTerminalAvailableKeyboardMappings$(props: { terminalId: String }): Rx.Observable<[] | void> {
...
}
Another flow parse error.

**Comment by vilchik-elena** [comment]
@marcdonovan this last one fails due to empty tuple, its in the list already

**Comment by NilSet** [comment]
Related to one of the above but possibly distinct, is the re-export syntax for types.
```
export type {foo} from 'bar';
```

**Comment by vilchik-elena** [comment]
@NilSet thanks, added to the list

**Comment by brunofourcroy** [comment]
+1 for `Rest property in object type, interface` !

**Comment by Stephen2** [comment]
Not sure if this is `Rest property in object type, interface`?

```
15: export type SizedSvgProps = {\|
--
  | 16:   ...SvgProps,
  | ^
  | 17:   +height?: number,
  | 18:   +width?: number,
  | 19: \|};
```

**Comment by Stephen2** [comment]
Basically object spread operator to extend a type into a superset of that type, with new properties

**Comment by kaiyoma** [comment]
+1 for the spread operator and `%checks`.

**Comment by vilchik-elena** [comment]
I'm closing this ticket as won't fix.
Currently we are gradually migrating SonarJS from our own parser to espree (ESLint parser), so it's unlikely we will fix this ticket in our parser.

Note that SonarJS espree-based part should already support all Flow syntaxes (as we are using babel parser for that), that means that some rules (more and more should come, see [MMF-1521](https://jira.sonarsource.com/browse/MMF-1521)) will be executed on your Flow files

**Comment by kaiyoma** [comment]
@vilchik-elena What steps do I need to take to get the new parsing? I've tried compiling SonarJS from source and uploading that JAR file to our SQ server, but I'm still seeing the same Flow parsing errors.

**Comment by vilchik-elena** [comment]
@kaiyoma during migration phase we keep both parsers: our own (failing to parse some flow syntax) and community (babel, which normally should parse everything in flow). Even if you see parsing error in logs, second parser will still parse it and report errors if any (unfortunately yet there are not so many rules relying on second parser). Also metrics and highlighting rely on first parser so your files will not have this data in SQ UI.

**Comment by kaiyoma** [comment]
@vilchik-elena Is there any way for us to track the progress your migration from the old parser to the new parser? After you fully migrate to the new parser, I'd like to rebuild SonarJS from source and install it in our SQ installation.

**Comment by vilchik-elena** [comment]
@kaiyoma This is the big ticket https://jira.sonarsource.com/browse/MMF-1409, you can follow it. I'm afraid it's not our priority now, we are not working on it now.

**Comment by lukeapage** [comment]
@vilchik-elena espree doesn't support flow. It doesn't support typescript - see https://github.com/eslint/espree/issues/278#issuecomment-223441850
You would need to use the babel-eslint project to parse with babel and provide that to eslint.

We have around 60/8000 of our files not parsing - this fails the azure devops task and prevents results being uploaded - so that means we cannot use sonar qube at all? Unless we exclude each of those 60 files and exclude files as they happen to get syntax in them that sonar doesn't understand?

Is there a way to ignore parse errors to unblock projects using sonar qube? I can't find a way.

Failing that is it possible to exclude javascript files from analysis but still get the benefit of code coverage analysis?

**Comment by vilchik-elena** [comment]
@lukeapage I'm afraid entirely excluding that files is the only workaround for you currently. 

FYI we already use [babel-eslint](https://github.com/SonarSource/SonarJS/blob/master/eslint-bridge/src/parser.ts#L47) to parse flow code when running ESLint-based rules. Still there are many more rules still relying on our java-based frontend. That's why you get parsing error logs. 

I'm not sure how it works, but you could consider configuring your azure devops task the way that it does not fail when `error` log.

**Comment by lukeapage** [comment]
@vilchik-elena I can make the build pass and it can run the sonar quality gate after analysis but it seems that inside the sonar qube analysis, when it gets parse errors, it does not generate any results, so nothing is uploaded to sonar qube.
My best current work around would be to make a script that modifies our source files to not have tokens in that sonar does not understand (e.g. `type abc { ...` as spread I think is the main issue) before running the analysis.
It seems crazy to me that this isn't a priority - JavaScript is moving fast.. is Sonar still working on their own parser in order to keep up with the standard? e.g. optional chaining may become part of the language by the end of the year, object spread officially got accepted etc.
At the very least, adding a option like you have for c++ to ignore parse errors would be great.

**Comment by lukeapage** [comment]
heres some code that might help someone else.. we run it and it removes the syntax that sonarjs doesn't understand.

```
const globby = require('globby');
const _ = require('lodash');
const fs = require('fs').promises;
const babelParser = require('@babel/parser');
const { traverse } = require('@babel/core');
const MagicString = require('magic-string');
const pLimit = require('p-limit');
const os = require('os');

(async () => {
    const files = await globby(['src/**/*.js']);

    const limit = pLimit(os.cpus().length);

    await Promise.all(
        _.map(files, (file) =>
            limit(async () => {
                try {
                    const contents = await fs.readFile(file, 'utf8');

                    const ast = babelParser.parse(contents, {
                        ranges: true,
                        sourceType: 'module',
                        plugins: [
                            'jsx',
                            'flow',
                            'classProperties',
                            'optionalChaining',
                            'nullishCoalescingOperator',
                            'dynamicImport',
                        ],
                    });

                    const newContents = new MagicString(contents);
                    let inFlowTypeNode = null;
                    traverse(
                        ast,
                        {
                            noScope: true,
                            enter(path) {
                                const node = path.node;

                                if (!inFlowTypeNode) {
                                    if (node.type === 'OptionalMemberExpression') {
                                        // optional chaining
                                        const questionMarkIndex = contents.indexOf(
                                            '?.',
                                            node.object.range[1],
                                        );
                                        newContents.overwrite(
                                            questionMarkIndex,
                                            questionMarkIndex + 1,
                                            '/*?*/',
                                        );
                                    } else if (
                                        node.type === 'LogicalExpression' &&
                                        node.operator === '??'
                                    ) {
                                        // nullish coalescing
                                        const questionMarkIndex = contents.indexOf(
                                            '??',
                                            node.left.range[1],
                                        );
                                        newContents.overwrite(
                                            questionMarkIndex,
                                            questionMarkIndex + 2,
                                            '/*??*/||',
                                        );
                                    } else if (
                                        node.type === 'Identifier' &&
                                        node.optional
                                    ) {
                                        // flow optional identifier
                                        newContents.overwrite(
                                            node.range[0] + node.name.length,
                                            node.range[0] + node.name.length + 1,
                                            '/*?*/',
                                        );
                                    } else if (
                                        node.type === 'TypeAnnotation' ||
                                        node.type === 'TypeAlias' ||
                                        node.type === 'TypeParameterInstantiation'
                                    ) {
                                        // Flow typings..

                                        let range = node.range;

                                        // if a type is exported, comment out the export too
                                        if (
                                            path.parentPath.node.type ===
                                            'ExportNamedDeclaration'
                                        ) {
                                            range = path.parentPath.node.range;
                                        }

                                        //  () : {
                                        //   } => {}
                                        // will cause asi to trigger on line 1, so if the character before
                                        // the type is a end parenthesis, extend it to include the new comment
                                        if (contents[range[0] - 1] === ')') {
                                            newContents.overwrite(
                                                range[0] - 1,
                                                range[0],
                                                '',
                                            );
                                            newContents.appendRight(range[1], ')');
                                        }

                                        // start the new comment
                                        newContents.appendRight(range[0], '/*');

                                        // uncomment block comments inside the range
                                        for (let i = 0; i < ast.comments.length; i++) {
                                            const comment = ast.comments[i];
                                            if (comment.type !== 'CommentBlock') {
                                                continue;
                                            }
                                            if (
                                                comment.start > range[0] &&
                                                comment.end < range[1]
                                            ) {
                                                newContents.overwrite(
                                                    comment.start,
                                                    comment.start + 1,
                                                    '*',
                                                );
                                                newContents.overwrite(
                                                    comment.end - 1,
                                                    comment.end,
                                                    '*',
                                                );
                                            }
                                        }

                                        // add the end comment
                                        newContents.appendLeft(range[1], '*/');
                                        inFlowTypeNode = node;
                                    }
                                }
                            },
                            exit(path) {
                                const node = path.node;
                                if (node === inFlowTypeNode) {
                                    inFlowTypeNode = null;
                                }
                            },
                        },
                        null,
                        { source: contents },
                    );

                    await fs.writeFile(file, newContents.toString(), 'utf8');
                } catch (e) {
                    console.log(e);
                    console.log('failure during processing of', file);
                }
            }),
        ),
    );
})();

```
